plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/BaseEqualsVisitor.java

```diff
@@ -53,19 +67,156 @@ public void visitMethodReferenceExpression(PsiMethodReferenceExpression expressi
   public void visitMethodCallExpression(@NotNull PsiMethodCallExpression expression) {
     super.visitMethodCallExpression(expression);
     final PsiExpression[] arguments = expression.getArgumentList().getExpressions();
-    final PsiExpression expression1;
-    final PsiExpression expression2;
     if (OBJECT_EQUALS.test(expression)) {
-      expression1 = PsiUtil.skipParenthesizedExprDown(ExpressionUtils.getEffectiveQualifier(expression.getMethodExpression()));
-      expression2 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
+      PsiExpression expression1 =
+        PsiUtil.skipParenthesizedExprDown(ExpressionUtils.getEffectiveQualifier(expression.getMethodExpression()));
+      PsiExpression expression2 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
+      checkTypes(expression, expression1, expression2);
     }
     else if (STATIC_EQUALS.test(expression)) {
-      expression1 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
-      expression2 = PsiUtil.skipParenthesizedExprDown(arguments[1]);
+      PsiExpression expression1 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
+      PsiExpression expression2 = PsiUtil.skipParenthesizedExprDown(arguments[1]);
+      checkTypes(expression, expression1, expression2);
     }
-    else {
-      return;
+    else if (PREDICATE_IS_EQUAL.test(expression)) {
+      PsiExpression expression1 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
+      PsiType psiType1 = getType(expression1);
+      PsiType psiType2 = resolveIsEqualPredicateType(expression);
+      if (psiType1 == null || psiType2 == null) {
+        return;
+      }
+      checkTypes(expression.getMethodExpression(), psiType1, psiType2);
+    }
+  }
+
+  @Nullable
+  private static PsiType resolveIsEqualPredicateType(PsiMethodCallExpression expression) {
+    PsiExpression upperPredicate = expression;
+    PsiElement parent = upperPredicate.getParent();
+    int max = 100;
+    int currentLevel = 0;
+    while (currentLevel <= max) {
+      currentLevel++;
+      if (parent instanceof PsiParenthesizedExpression || parent instanceof PsiReferenceExpression) {
```

Comment:
`parent instanceof PsiReferenceExpression` is confusing here. As I understand, you are doing this solely for `.negate()` support. First, I would use a separate CallMatcher for `negate()`, as `parent instanceof PsiExpressionList && isPredicateSourceForIsEqual(parent.getParent())` test below is irrelevant for `negate` case. Second, I would use something like:
```java
var call = ExpressionUtils.getCallForQualifier(parent);
if (PREDICATE_NEGATE.test(call)) {
  parent = call;
}
```

Reply:
Thank you. Try to simplify logic. Could you inspect?